File: dp/coins.js
// https://www.geeksforgeeks.org/coin-change-dp-7/
// recursion

// Input: sum = 4, coins[] = {1,2,3}, 
// Output: 4
// Explanation: there are four solutions: {1, 1, 1, 1}, {1, 1, 2}, {2, 2}, {1, 3}. 

var coinChange = function (sum, coins) {
    console.log(sum, coins)
    if (sum === 0) return 1;
    if (sum < 0) return 0;
    if (coins.length === 0) return 0;
    return coinChange(sum - coins[0], coins) + coinChange(sum, coins.slice(1));
}
console.log(coinChange(4, [1, 2, 3]))

File: dp/coins.py
# the problem does

def coinChange(sum, coins):
    dp = [0 for i in range(sum+1)]
    dp[0] = 1
    for i in range(len(coins)):
        for j in range(coins[i], sum+1):
            dp[j] += dp[j - coins[i]]
    return dp[sum]


print(coinChange(5, [1, 2, 5]))
print(coinChange(4, [1, 2, 3]))
print(coinChange(3, [2]))
